xds/internal/balancer/cdsbalancer/cluster_handler.go

```diff
@@ -176,62 +162,74 @@ func (c *clusterNode) handleResp(clusterUpdate xdsclient.ClusterUpdate, err erro
-	var deletedChild bool
-	// deltaInChildUpdateFields determines whether there was a delta in the clusterUpdate fields (forgetting the children).
-	// This will be used to help determine whether to pingClusterHandler at the end of this callback or not.
-	deltaInChildUpdateFields := deltaInClusterUpdateFields(clusterUpdate, c.clusterUpdate)
+	var createdChild, deletedChild bool
 
-	c.receivedUpdate = true
-	c.clusterUpdate = clusterUpdate
+	// This map will represent the current children of the cluster. It will be first added to in order to represent the
+	// new children. It will then have any children deleted that are no longer present. Then, from the cluster update received,
+	// will be used to construct the new child list.
+	mapCurrentChildren := make(map[string]*clusterNode)
+	for _, child := range c.children {
+		mapCurrentChildren[child.clusterUpdate.ClusterName] = child
+	}
 
-	// This map will be empty if the cluster update specifies cluster is an EDS or LogicalDNS cluster, as will have no children.
-	newChildren := make(map[string]struct{})
-	if clusterUpdate.ClusterType == xdsclient.ClusterTypeAggregate {
-		for _, childName := range clusterUpdate.PrioritizedClusterNames {
-			newChildren[childName] = struct{}{}
+	// Add and construct any new child nodes.
+	for child := range newChildren {
+		if _, inChildrenAlready := mapCurrentChildren[child]; !inChildrenAlready {
+			createdChild = true
+			mapCurrentChildren[child] = createClusterNode(child, c.clusterHandler.xdsClient, c.clusterHandler)
 		}
 	}
 
-	for _, child := range c.children {
-		// If the child is still present in the update, then there is nothing to do for that child name in the update.
-		if _, found := newChildren[child.clusterUpdate.ClusterName]; found {
-			delete(newChildren, child.clusterUpdate.ClusterName)
-		} else { // If the child is no longer present in the update, that cluster can be deleted.
+	// Delete any child nodes no longer in the aggregate cluster's children.
+	for child := range mapCurrentChildren {
+		if _, stillAChild := newChildren[child]; !stillAChild {
 			deletedChild = true
-			child.delete()
-			// There is no need to delete the pointer from the current children list. A new list of pointers will be constructed
-			// at the end of this callback to replace the list of pointers.
+			mapCurrentChildren[child].delete()
+			delete(mapCurrentChildren, child)
 		}
 	}
 
-	// Whatever clusters are left over here from the update are all new children, so create CDS watches for those clusters.
 	// The order of the children list matters, so use the clusterUpdate from xdsclient as the ordering, and use that logical
-	// ordering for the new children list. This will be a mixture of child nodes already constructed and also children
-	// that need to be created.
+	// ordering for the new children list. This will be a mixture of child nodes which are all already constructed in the
+	// mapCurrentChildrenMap.
 	var children []*clusterNode
```

Comment:
Nit: you can make a slice with the exact length you need. Otherwise it's going to double every time the limit is reached.

`children = make([]*clusterName, len(clusterUpdate.PrioritizedClusterNames), 0)`

Reply:
Interesting just like arraylists in Java. Will preallocate.

Reply:
Changed.

Reply:
Why cap it at 0? I don't understand this comment.

Reply:
I switched the two, make(..., 0, len(...)) and it worked.